File: InMemoryCompilers/Ruby/jsonrepl.rb
require 'json'
require 'stringio'

trap 'INT' do exit end
$stdout.sync = true
$realStdout = $stdout
    
def resp(result)
    result["backendVersion"] = "one:ruby:jsonrepl:20180122"
    $realStdout.puts(JSON.generate(result))
end

while requestLine = gets
    begin
        $stdout = StringIO.new
        request = JSON.parse(requestLine)
        fileNames = request["packageSources"].map{|x| x["fileName"].sub(".rb", "")}

        def evalCode(code, fileNames)
            for fn in fileNames do
                code = code.sub("require '#{fn}'", "")
            end
            eval code
        end

        evalCode(request['stdlibCode'], fileNames) if request['stdlibCode']
        for pkgSource in request["packageSources"] do
            evalCode(pkgSource["code"], fileNames)
        end
        result = evalCode(request['code'], fileNames)
        resp({ :result => $stdout.string })
    rescue Exception
        resp({ :exceptionText => "#{$@.first}: #{$!.message} (#{$!.class})" + $@.drop(1).map{|s| "\n\t#{s}"}.join("") })
    ensure
        $stdout = $realStdout
    end
end

File: InMemoryCompilers/Python/jsonrepl.py
#!/usr/bin/env python
import json
import traceback
import sys
import imp
from cStringIO import StringIO

def resp(result):
    result["backendVersion"] = "one:python:jsonrepl:20180122"
    print json.dumps(result)

while True:
    try:
        requestLine = raw_input()
    except:
        break

    try:
        request = json.loads(requestLine)
        if request["cmd"] == "compile":
            original_stdout = sys.stdout
            sys.stdout = result_stdout = StringIO()
            try:
                for pkgSource in request["packageSources"]:
                    name = pkgSource["fileName"].replace(".py", "")
                    sys.modules[name] = imp.new_module(name)
                    exec(pkgSource["code"], sys.modules[name].__dict__)
                exec(request["code"], {})
            finally:
                sys.stdout = original_stdout

            resp({ "result": result_stdout.getvalue() })
        else:
            resp({ "errorCode": "unknown_command", "exceptionText": "Unknown command: " + request["cmd"] })
    except Exception as e:
        resp({ "errorCode": "unexpected_exception", "exceptionText": traceback.format_exc() })
